src/Autofac/Core/Activators/Reflection/BoundConstructor.cs

```diff
@@ -0,0 +1,131 @@
+﻿// This software is part of the Autofac IoC container
+// Copyright © 2011 Autofac Contributors
+// https://autofac.org
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Reflection;
+
+namespace Autofac.Core.Activators.Reflection
+{
+    /// <summary>
+    /// Represents the outcome of a single bind attempt by a <see cref="ConstructorBinder"/>.
+    /// </summary>
+    [SuppressMessage(
+        "Performance",
+        "CA1815:Override equals and operator equals on value types",
+        Justification = "Comparison of two BoundConstructor instances has no meaning.")]
+    public class BoundConstructor
+    {
+        private readonly Func<object?[], object>? _factory;
+        private readonly Func<object?>[]? _valueRetrievers;
+        private readonly ParameterInfo? _firstNonBindableParameter;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BoundConstructor"/> class for a successful bind.
+        /// </summary>
+        /// <param name="constructor">The constructor.</param>
+        /// <param name="factory">The instance factory.</param>
+        /// <param name="valueRetrievers">The set of value-retrieval functions.</param>
+        public BoundConstructor(ConstructorInfo constructor, Func<object?[], object> factory, Func<object?>[] valueRetrievers)
```

Comment:
The constructor "interface" here seems somewhat unintuitive. One constructor is for "successful" binds, one is for "unsuccessful" binds.

Would it be more intuitive to have static factory methods where the constructors are _internal_ and then you do something like:

`BoundConstructor.ForUnsuccessfulBind(info, parameter);`

or

`BoundConstructor.CanNotInstantiate(info, parameter);`

or something like that? To make it more clear about what you're getting.

Reply:
You're right, it's not clear; I'll use the static factory approach.